pkg/util/validation/limits.go

```diff
@@ -778,8 +782,32 @@ func (o *Overrides) IgnoreOOOExemplars(userID string) bool {
 	return o.getOverridesForUser(userID).IgnoreOOOExemplars
 }
 
+// ActiveSeriesCustomTrackersConfig returns all active series custom trackers that should be used for
+// the input tenant. The trackers are the merge of the configure base and additional custom trackers.
 func (o *Overrides) ActiveSeriesCustomTrackersConfig(userID string) asmodel.CustomTrackersConfig {
-	return o.getOverridesForUser(userID).ActiveSeriesCustomTrackersConfig
+	// Check if we already have a cached version of the merged ones.
+	if merged := o.getOverridesForUser(userID).activeSeriesMergedCustomTrackersConfig; merged != nil {
+		return *merged
+	}
+
+	// Merge the base trackers with the additional ones.
+	merged := asmodel.MergeCustomTrackersConfig(
+		o.activeSeriesBaseCustomTrackersConfig(userID),
+		o.activeSeriesAdditionalCustomTrackersConfig(userID),
+	)
+
+	// Cache it.
+	o.getOverridesForUser(userID).activeSeriesMergedCustomTrackersConfig = &merged
```

Comment:
This looks racy 🤔 Shouldn't we coordinate the access here and above where we're reading it?

Reply:
You're right, it's racy and I missed it. I pushed an option to fix the race in https://github.com/grafana/mimir/pull/10428/commits/877b9ab0f2f47114933af7e42ddac52eaa608583